File: src/core/transaction.rs
use super::amount::Amount;
use super::coin::Coin;
use super::key::{WalletKeyPair, WalletPub};
use super::signature::Signature;

pub struct Transaction<'a> {
    pub amount: Amount,
    pub coin: Coin,
    pub from_keypair: &'a WalletKeyPair,
    pub to_addr: &'a WalletPub,
    pub timestamp_millis: i64,
}

pub struct SignedTransaction<'a> {
    pub transaction: Transaction<'a>,
    pub signature: Signature,
}

impl<'a> Transaction<'a> {
    pub fn new(
        amount: Amount,
        coin: Coin,
        from: &'a WalletKeyPair,
        to: &'a WalletPub,
        timestamp_millis: i64)
        -> Self {

        Transaction {
            amount: amount,
            coin: coin,
            from_keypair: from,
            to_addr: to,
            timestamp_millis: timestamp_millis,
        }
    }

    pub fn serialized_for_signing(&self) -> Vec<u8> {
        let a = self.amount.serialize();
        let c = self.coin.serialize();
        let f = self.from_keypair.public_key().serialize();
        let t = self.to_addr.serialize();
        let ts = format!("{}_unix_millis", self.timestamp_millis);
        let fmt = format!(
            "amount:{},coin:{},from:{},to:{},timestamp:{}",
            a, c, f, t, ts);
        println!("{}", fmt);
        fmt.into_bytes()
    }
}

impl<'a> SignedTransaction<'a> {
    pub fn sign(transaction: Transaction<'a>) -> Self {
        let sig = Signature::sign(
            &transaction.from_keypair,
            &(transaction.serialized_for_signing()));

        SignedTransaction {
            transaction: transaction,
            signature: sig,
        }
    }

    pub fn signature_is_valid(&self) -> bool {
        self.signature.msg_has_valid_sig(
            &self.transaction.from_keypair.public_key(),
            &self.transaction.serialized_for_signing())
    }

    pub fn serialized_for_block(&self) -> Vec<u8> {
        vec![1, 2, 3, 4]
    }
}

File: core/transaction.py
from core.amount import Amount
from core.key_pair import Address, KeyPair
from core.serializable import Serializable, Ser
from core.signature import Signature
from core.timestamp import Timestamp
from typing import Union

class Transaction(Serializable):
    def __init__(
            self,
            amount: Amount,
            timestamp: Timestamp,
            from_addr: Union[Address, None],
            to_addr: Address) -> None:
        self.amount = amount
        self.timestamp = timestamp
        self.from_addr = from_addr
        self.to_addr = to_addr

    def serializable(self) -> Ser:
        from_addr = self.from_addr.serializable() if self.from_addr else None,
        return {
            "amount": self.amount.serializable(),
            "timestamp": self.timestamp.serializable(),
            "from_addr": from_addr,
            "to_addr": self.to_addr.serializable(),
        }

class SignedTransaction(Serializable):
    def __init__(self, transaction: Transaction, signature: Signature) -> None:
        self.transaction = transaction
        self.signature = signature

    @staticmethod
    def sign(transaction: Transaction, key_pair: KeyPair) -> "SignedTransaction":
        sig = key_pair.sign(transaction.serialize())
        return SignedTransaction(transaction, sig)

    def is_valid(self) -> bool:
        if self.transaction.from_addr is not None:
            verify_addr = self.transaction.from_addr
        else:
            verify_addr = self.transaction.to_addr

        return verify_addr.signature_is_valid(
                self.transaction.serialize(),
                self.signature)
    
    def serializable(self) -> Ser:
        return {
            "transaction": self.transaction.serializable(),
            "signature": self.signature.serializable(),
        }
